src/blog/0.3/linear-types-borrowing.vmd

```diff
@@ -0,0 +1,673 @@
+ 1. We can't use certain patterns that decouple our code and make it more flexible, like [observers](https://www.reddit.com/r/rust/comments/pwqju6/is_there_an_underlying_reason_that_idiomatic_rust/), [back-references](https://users.rust-lang.org/t/back-reference-to-the-parent-struct/7413/2), [dependency references](https://en.wikipedia.org/wiki/Dependency_injection), [callbacks](https://news.ycombinator.com/item?id=12029238), and [delegates](https://medium.com/@nimjea/delegation-pattern-in-swift-4-2-f6aca61f4bf5).
+
+
+All of the above drawbacks exist because neither linear typing nor borrow checking can have *shared mutability*. In other words, multiple mutable references to the same data.
+
+
+<slice/>
+
+# Addressing the Drawbacks
+
+Are there any ways we can address those downsides without sacrificing the benefits?
+
+
+In other words, is there a way we can have linear types _and_ shared mutability?
+
+
+Adding borrow checking partially helps with drawback #1 [# Partially because borrow checking lets us have temporary references, not persistent ones. The borrow checker has a hard time putting references into data structures.] but it still has drawbacks #2, #3, and #4. Rust's `Rc` and `RefCell` helps with those, but they're incompatible with linear types' benefits [# We can't put a linear type into an `Rc` because `Rc` throws away its contents when the last alias disappears.], and have their own drawbacks (like forcing heap allocation and not having [higher RAII](https://verdagon.dev/blog/higher-raii-7drl)).
+
+
+[# E: How is this next part, where we blend gen refs back in? I tried to find an elegant way to build up to "the solution was with us all along".]
+
+The other solution, surprisingly, is to blend generational references back in.
+
+Each object is still owned linearly; only one place can own an object at a time. But we'll also allow a generational reference to point at the object.
+
+With this, we can address a lot of the drawbacks:
+
+ 1. We can have as many references to data as we want, though there is still only one owner.
+ 1. We don't need to restructure our program to accommodate them.
+ 1. We can use patterns like intrusive data structures and graphs. [# In Vale, we would still have all nodes owned by a central collection, but we could still have references between nodes just fine.]
+ 1. We can use observers, back-references, dependency references, callbacks, delegates, whatever we want.
+ 1. We still don't have data races, because Vale changes the generations of any objects that cross region boundaries (unless we safely avoid that with [isolates](/blog/zero-cost-borrowing-regions-part-2-isolates)).
+
+
+Generational references seem to occupy a sweet spot because they allow objects to be linear (they're owned by only one place), yet allow shared mutability in a way that doesn't artificially extend the lifetime of the object.
```

Comment:
"in a way that doesn't artificially extend the lifetime of the object."
Where did this come from? First time you've mentioned it... maybe add "like blah blah blah does" or something